qiskit/primitives/containers/sampler_pub_result.py

```diff
@@ -0,0 +1,88 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Sampler Pub result class
+"""
+
+from __future__ import annotations
+
+from .bit_array import BitArray
+from .pub_result import PubResult
+
+
+class SamplerPubResult(PubResult):
```

Comment:
I will summarize my understanding of the offline discussion we had:

 - Having methods with the same name on a different but related class that do something slightly different may be confusing. That is, `job.result()[0].get_counts()`.
 - It may be nice, for these methods, to have the ability to contatenate data along the bits axis.
 - There is concern about heterogeneous data types. At the moment, the only problem would be between IQ data (which is a np.complex64 array) and `BitArray`. In the future, we may have other types due to a richer type structure in `QuantumCircuit`.

For these reasons, we discussed replacing the three nice methods you have here with an attribute or method that joins data of a particular type.

For an attribute, this might look something like:
```python
    @property
    def joint_bit_array(self) -> BitArray:
        """Join data from all registers with BitArray data into a single BitArray.
        
        This method maintains register order specified by the original circuit.
        """
        <implementation>
```

A method might let you do something generic like
```python
    def joint_data(self, names=Iterable[str] | None = None, data_type: T = BitArray) -> T:
        """Join data from provided (defaulting to all) registers of the given type."""
```

@ElePT your advice would be appreciated :)

Reply:
Thanks all. It seems like the method form is preferred, and I like that better too. @t-imamichi would something like the following be possible?

```python
def join_data(self, names: Iterable[str] | None = None) -> BitArray | ndarray:
    """Join data from many registers into one data container.

    Data is joined along the bits axis. For example, for :class:`~.BitArray` data, this corresponds
    to bitstring concatenation.

    Args:
        TODO: check that the behaviour described below matches how Result joins data!
        names: Which registers to join. Their order is maintained, for example, given
            ``["alpha", "beta"]``, the data from register ``alpha`` is placed to the left of the 
            data from register ``beta``. When ``None`` is given, this value is set to the 
            ordered list of register names, which will have been preserved from the input circuit
            order.

    Returns:
        Joint data.

    Raises:
        TypeError: If specified data comes from incompatible types.
    """
    names = list(self.data if names is None else names)
    ...
```

@ElePT I'll think a bit more about your shortcut idea and ask around.